### PR TITLE
Tracing Infrastructure

### LINKED ISSUE #289: API tracing
API tracing is a feature that allows to serialize the workload that the library is processing over a period of time, so that it can be replayed later, potentially on a different machine and with a different GPU backend. It opens up a way to capture bugs (for investigating either the implementation or the application issues), share interesting workloads (as a form of art), or even do benchmarking.

At this point, it's clear that WebGPU as an API needs some tooling for workload capturing. There probably will appear some JS-based tools developed by the users and browser vendors. This issue is focuses on the native targets. Let's see what options we have to proceed:

## A: Intercept in Rust, serialize with Serde.

This basically means transforming all the workload inputs (object descriptors, handles, etc), into pure Rust structs that implement Serde traits and recording them.

What this gives us:
  - Convenience of Rust and its ecosystem (i.e can use RON format as destination).
  - Can start the work today, not blocked by anything
  - Can build an internal testing and benchmarking suite based on the captured format
  - If used in conjunction with #72, could support recording from a particular point (as opposed to the start of an application).

What's not clear is where this hookup should live. It could be a separate layer on top of `wgpu-native`, it could be a part of `wgpu-remote`, or be a part of `wgpu-rs` even.

## B: develop a backend for RenderDoc and integrate with it

Given that RenderDoc became a central piece of the developer graphics tools, and it's open source, there is an opportunity here to start collaborating.

What this gives us:
  - RenderDoc is not an afterthought, we get it from day 1
  - convenience of the GUI

## C: wait for the native tooling to emerge

We are not the only party interested in this technology. Obviously, Google is either already in process, or will attempt to build some tooling for tracing the API at the C level. Something like `wgpuTrace` as an analogy to `vkTrace` and `apiTrace`.

For us this plan means letting the things to advance by Google and assisting where time allows, focusing on interoperability with wgpu-native implementation (since clearly a hypothetical `wgpuTrace` will be catering to Dawn first and foremost).

What this gives us:
  - path of least resistance

---

Now, I'm personally leaning towards (A) since I don't think this is going to be a huge amount of work, I think it could play well with #72, it will allow us to inspect and modify captures in RON by hands, and build some testing/benchmarking without relying on external suites.

My understanding is that when/if `wgpuTrace` becomes a thing, we'll just set up a tooling for bidirectional conversion to its binary format, effectively mimicking how ApiTrace trace is different from the dump.

This is a heavy RFC, please share your thoughts!

### PR DESCRIPTION
**Pull request by kvark**
Closes #289
Downstream changes: https://github.com/gfx-rs/wgpu-rs/pull/291 https://github.com/gfx-rs/wgpu-native/pull/17
Note: close to half of the new LOCs belong to `Cargo.lock`, which we are now including since we have a binary target crate `player`.

## Why do we need that?

Have you ever seen a case that doesn't work for somebody? Often, reproducing the complete setup requires elaborate toolchains, external resources, and closed-source code. Even with all of that, the difference between platforms may hide the issue. [RenderDoc](https://renderdoc.org/) or Metal captures do not help, since they are extremely driver/hardware-dependent.

This problem is amplified for WebGPU implementation in Gecko. Some workloads are rendered incorrectly, and debugging them within the browser becomes too much of a challenge.

Temporary limitations (can be fixed later):
  - replaying has to happen on the same backend. This is easy to work around by just mass-replacing the backend in the trace (i.e. `s/Vulkan/Metal/g`)
  - buffer readbacks are not recorded

Tracing allows us to swiftly isolate and reproduce rendering issues. We can now consider a [Warden](https://github.com/gfx-rs/gfx/pull/1589)-like testing and benchmarking framework built around the traces.

## What the hell it is?

This is a state of art tracing infrastructure built into `wgpu-core`. It traces all the commands and resource changes, very similarly to `apitrace`. Tracing code is enabled by a "trace" feature, and the actual support is enabled at the device instantiation, controlled at run-time. There is no run-time overhead for running with tracing feature but in disabled state.

The result of tracing is a directory, containing the following files:
  - "trace.ron" is basically a list of all the actions taken by the app, in [RON](https://github.com/ron-rs/ron) format.
  - each buffer update, or a shader module created, is placed into a separate binary file referenced by the trace.

The new "player" application can be used to replay the traces. It's launched as:
```rust
target/debug/player <trace-dir>
```
When built with "winit" feature, it's able to replay the workloads that operate on a swapchain. It renders each frame consequently, then waits for the user to close the window. When built without "winit", it launches in console mode and can replay any trace that doesn't use swapchains.

## How does it look?

Having the trace in RON makes it easy to inspect visually (by humans!). There is no separate "dump" phase. Each submission is labelled, and the trace is easy to navigate. Moreover, everything is human-editable: you can tweak it to see what changes would be needed in an application in order to fix a rendering issue. For example, you can change the blend modes and re-play.

```ron
CreateBindGroupLayout(
    id: Id(0, 1, Metal),
    label: "",
    entries: [
        (
            binding: 0,
            visibility: (
                bits: 4,
            ),
            ty: StorageBuffer,
            multisampled: false,
            has_dynamic_offset: false,
            view_dimension: D2,
            texture_component_type: Float,
            storage_texture_format: Rgb10a2Unorm,
        ),
    ],
),
CreateBindGroup(
    id: Id(0, 1, Metal),
    label: "",
    layout_id: Id(0, 1, Metal),
    entries: {
        0: Buffer(
            id: Id(1, 1, Metal),
            offset: 0,
            size: 16,
        ),
    },
),
CreatePipelineLayout(
    id: Id(0, 1, Metal),
    bind_group_layouts: [
        Id(0, 1, Metal),
    ],
),
CreateComputePipeline(
    id: Id(0, 1, Metal),
    desc: (
        layout: Id(0, 1, Metal),
        compute_stage: (
            module: Id(0, 1, Metal),
            entry_point: "main",
        ),
    ),
),
Submit(2, [
    CopyBufferToBuffer(
        src: Id(0, 1, Metal),
        src_offset: 0,
        dst: Id(1, 1, Metal),
        dst_offset: 0,
        size: 16,
    ),
    RunComputePass(
        commands: [
            SetPipeline(Id(0, 1, Metal)),
            SetBindGroup(
                index: 0,
                num_dynamic_offsets: 0,
                bind_group_id: Id(0, 1, Metal),
            ),
            Dispatch((4, 1, 1)),
        ],
        dynamic_offsets: [],
    ),
    CopyBufferToBuffer(
        src: Id(1, 1, Metal),
        src_offset: 0,
        dst: Id(0, 1, Metal),
        dst_offset: 0,
        size: 16,
    ),
]),
```

## Alternatives considered

Q: Why not just add support to `apitrace` (or `renderdoc`) instead?
A: There is a benefit of having this Rusty: we can use heavy lifters like Serde and RON, we use pattern matching extensively. We can still consider something more independent of the language in the future, something with C API that would work for both Dawn and wgpu-native. I just don't want to invest in that now, and going for the local solution was quite straightforward.

Q: Why not slice the execution at a particular time, instead of recording everything from the beginning?
A: I think both approaches are needed, but for different purposes. Tracing allows to share and reproduce rendering issues. Also can aid in the test corpus building. Slicing vertically would help to investigate internal state problems. We can still get that in the future.

Q: Why RON and not JSON/YAML/etc?
A: Experience with WebRender proved RON to be ideal for this kind of use. Technically, there are no blockers from writing to another format - it's a matter of a single line change, plus some extra dependencies. After all, this is all going through Serde :)

### CODE REVIEW

**Comment by aloucks** [comment]
Have you considered using `json` or `yaml` for serialization instead of `ron`? It might be easier in the long run to incorporate other tooling if you're using a well-supported standard serialization format.

**Comment by kvark** [comment]
@aloucks did you read to the end? This is exactly the last item of "Alternatives considered" section, at the bottom of the description.
TL;DR: Switching this to any other Serde based format is a matter of a handful of lines, this code doesn't really depend on it in any way. If there is interest in larger interop, e.g. with Dawn using this very format, we can add YAML/JSON/whatever in a matter of minutes.

**Comment by aloucks** [comment]
@kvak Ah, I see that now 👍

File: wgpu-types/src/lib.rs [hunk 1]
```diff
@@ -4,13 +4,16 @@
 
 #[cfg(feature = "peek-poke")]
 use peek_poke::PeekPoke;
-#[cfg(feature = "serde")]
-use serde::{Deserialize, Serialize};
+#[cfg(feature = "replay")]
```
Comment (grovesNL) [inline wgpu-types/src/lib.rs:7]: I think it might be confusing for people looking to enable serialization to have to enable `trace` or `replay`. Most people would probably look for a `serde` or `serde1` feature because that's the common place to look.

For example, users may just want to serialize their types, but shouldn't necessarily need to understand the association between serialization and trace/replay (even if trace/replay happen to also require serialization).

File: player/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+use std::{
+    ffi::CString,
+    fmt::Debug,
+    fs::File,
+    marker::PhantomData,
+    path::{Path, PathBuf},
+    ptr,
+};
+
+macro_rules! gfx_select {
```
Comment (grovesNL) [inline player/src/main.rs:24]: Should we expose `gfx_select` from wgpu-core and use that?

File: wgpu-core/src/command/transfer.rs [hunk 3]
```diff
@@ -90,6 +96,18 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         // borrow the buffer tracker mutably...
         let mut barriers = Vec::new();
 
+        #[cfg(feature = "trace")]
+        match cmb.commands {
```
Comment (grovesNL) [inline wgpu-core/src/command/transfer.rs:100]: Looks like we could have `if let Some(ref mut list) = cmb.commands) { /* ... */ }` for these and some other places

File: wgpu-core/src/device/mod.rs [hunk 4]
```diff
@@ -204,7 +212,8 @@ impl<B: GfxBackend> Device<B> {
         mem_props: hal::adapter::MemoryProperties,
         non_coherent_atom_size: u64,
         supports_texture_d24_s8: bool,
-        max_bind_groups: u32,
+        desc: &wgt::DeviceDescriptor,
+        #[cfg(feature = "trace")] trace_path: Option<&std::path::Path>,
```
Comment (grovesNL) [inline wgpu-core/src/device/mod.rs:216]: Same here, I think it would be better to always have the argument and just pass `None` when `trace` isn't passed

File: wgpu-core/src/swap_chain.rs [hunk 5]
```diff
@@ -151,6 +156,15 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
             .texture_views
             .register_identity(view_id_in, view, &mut token);
 
+        #[cfg(feature = "trace")]
+        match device.trace {
```
[the remark this answers is not in the document]
Comment (kvark) [reply wgpu-core/src/swap_chain.rs:160]: I'd love to do that, but it's not possible today. Another alternative would be `trace.as_ref().map(|trace| ...)` but that may run into complications with borrowing into the closure... so I think for now `match` is the best of what we can do.

File: wgpu-types/src/lib.rs (hunk 1 above)
Comment (kvark) [reply wgpu-types/src/lib.rs:7]: Right, I agree. Also, I think there is value to separating those even outside of the trace/replay scope: quite often the user knows exactly which side is needed, and deriving both affects compile times quite a bit.
Perhaps, this is something to address by documentation?

File: player/src/main.rs (hunk 2 above)
Comment (kvark) [reply player/src/main.rs:24]: possibly, yes. Today, the one used in wgpu-rs is slightly different though. We may end up with small differences across the users, and the macro itself is pretty small. Let's keep an eye on it for follow-ups?
